[te] Introduce likely/unlikely CompareSelect hint

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51771 [te] Benchmark vml-based logit
* #51752 [te] Implement log approximation using the VML approach
* **#51751 [te] Introduce likely/unlikely CompareSelect hint**
* #51749 [te][llvm] Generate arithmetic vs logical right shift as appropriate

Similar in spirit to the `__builtin_expect` C intrinsic, it's useful
to be able to hint the expected branch direction in a tensor expression.  Using
this flag has a few effects on codegen:

- The CompareSelect is generated using conditional branches, rather than selects
- The conditional branches are strongly hinted (like, 100000:1) in the indicated direction
- A vectorized hinted CompareSelect computes its condition in parallel with a
  mask "reduction" (e.g. a bitcast from `<i1 x 8>` to `<i*>`).  In AVX terms
  this sequence might look like:
```
vpcmpgtd %ymm0, %ymm1, %ymm2
vmovmskps %ymm2, %eax
```

The motivating case for this addition is an attempt I'm making to replicate
fast transcendentals using tensor expressions.  Floating-point numbers have
lots of special cases (denormals, inf, nan) that need special handling, and
it's convenient to be able to punt that handling off to a slow path while
keeping the fast path nice and tight.

Differential Revision: [D26246401](https://our.internmc.facebook.com/intern/diff/D26246401/)

bertmaher committed:
[te] Introduce likely/unlikely CompareSelect hint

Similar in spirit to the `__builtin_expect` C intrinsic, it's useful
to be able to hint the expected branch direction in a tensor expression.  Using
this flag has a few effects on codegen:

- The CompareSelect is generated using conditional branches, rather than selects
- The conditional branches are strongly hinted (like, 100000:1) in the indicated direction
- A vectorized hinted CompareSelect computes its condition in parallel with a
  mask "reduction" (e.g. a bitcast from `<i1 x 8>` to `<i*>`).  In AVX terms
  this sequence might look like:
```
vpcmpgtd %ymm0, %ymm1, %ymm2
vmovmskps %ymm2, %eax
```

The motivating case for this addition is an attempt I'm making to replicate
fast transcendentals using tensor expressions.  Floating-point numbers have
lots of special cases (denormals, inf, nan) that need special handling, and
it's convenient to be able to punt that handling off to a slow path while
keeping the fast path nice and tight.

Differential Revision: [D26246401](https://our.internmc.facebook.com/intern/diff/D26246401/)

[ghstack-poisoned]

asuhan commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -698,53 +720,93 @@ void LLVMCodeGenImpl::visit(const Min* v) {
-      default:
-        // TODO: change to a proper error report
-        throw std::runtime_error("invalid operator type");
+  auto genUnbiased = [this, v]() -> llvm::Value* {
```
Do these two (`genUnbiased` and `genBiased`) really need to be lambdas, wouldn't they be better as methods?

bertmaher replied:
I kind of prefer lambdas here for the maybe slightly silly reason that they want to use `this`, but they're just local helpers so I hate adding them to the class signature  ¯\_(ツ)_/¯

asuhan commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -698,53 +720,93 @@ void LLVMCodeGenImpl::visit(const Min* v) {
+    auto then_block = llvm::BasicBlock::Create(getContext(), "then", fn_);
+    auto else_block = llvm::BasicBlock::Create(getContext(), "else", fn_);
+    auto end_block = llvm::BasicBlock::Create(getContext(), "block", fn_);
+    constexpr int32_t total_weight = 100000;
```
What's the significance of 100000 here, couldn't it be 10 or even 1?

bertmaher replied:
We talked about this offline, but just for posterity: this feature doesn't seem well documented.  I basically swiped this usage from the llvm repo itself, e.g. [this instrumentation in the ASAN implementation](https://github.com/llvm/llvm-project/blob/cfeecdf7b6df9cd89488948b440f8deeb458104c/llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp#L1677).

asuhan reviewed (approved):
Looks good, address the two comments if you agree.

bertmaher committed:
rebase on "[te] Introduce likely/unlikely CompareSelect hint"


Similar in spirit to the `__builtin_expect` C intrinsic, it's useful
to be able to hint the expected branch direction in a tensor expression.  Using
this flag has a few effects on codegen:

- The CompareSelect is generated using conditional branches, rather than selects
- The conditional branches are strongly hinted (like, 100000:1) in the indicated direction
- A vectorized hinted CompareSelect computes its condition in parallel with a
  mask "reduction" (e.g. a bitcast from `<i1 x 8>` to `<i*>`).  In AVX terms
  this sequence might look like:
```
vpcmpgtd %ymm0, %ymm1, %ymm2
vmovmskps %ymm2, %eax
```

The motivating case for this addition is an attempt I'm making to replicate
fast transcendentals using tensor expressions.  Floating-point numbers have
lots of special cases (denormals, inf, nan) that need special handling, and
it's convenient to be able to punt that handling off to a slow path while
keeping the fast path nice and tight.

Differential Revision: [D26246401](https://our.internmc.facebook.com/intern/diff/D26246401/)

[ghstack-poisoned]

bertmaher committed:
rebase on "[te] Introduce likely/unlikely CompareSelect hint"


Similar in spirit to the `__builtin_expect` C intrinsic, it's useful
to be able to hint the expected branch direction in a tensor expression.  Using
this flag has a few effects on codegen:

- The CompareSelect is generated using conditional branches, rather than selects
- The conditional branches are strongly hinted (like, 100000:1) in the indicated direction
- A vectorized hinted CompareSelect computes its condition in parallel with a
  mask "reduction" (e.g. a bitcast from `<i1 x 8>` to `<i*>`).  In AVX terms
  this sequence might look like:
```
vpcmpgtd %ymm0, %ymm1, %ymm2
vmovmskps %ymm2, %eax
```

The motivating case for this addition is an attempt I'm making to replicate
fast transcendentals using tensor expressions.  Floating-point numbers have
lots of special cases (denormals, inf, nan) that need special handling, and
it's convenient to be able to punt that handling off to a slow path while
keeping the fast path nice and tight.

Differential Revision: [D26246401](https://our.internmc.facebook.com/intern/diff/D26246401/)

[ghstack-poisoned]
